src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.TestTypes.cs

```diff
@@ -0,0 +1,1320 @@
+                {
+                    "type": ["object","null"],
+                    "properties": { "X": { "type": ["string","integer"] } }
+                }
+                """);
+
+            yield return new TestData<PocoWithCustomNumberHandlingOnProperties>(
+                Value: new() { X = 1, Y = 2, Z = 3 },
+                AdditionalValues: [
+                    new() { X = 1, Y = double.NaN, Z = 3 },
+                    new() { X = 1, Y = double.PositiveInfinity, Z = 3 },
+                    new() { X = 1, Y = double.NegativeInfinity, Z = 3 },
+                ],
+                ExpectedJsonSchema: """
+                {
+                  "type": ["object","null"],
+                  "properties": {
+                    "X": { "type": ["string", "integer"] },
+                    "Y": {
+                      "anyOf": [
+                        { "type": "number" },
+                        { "enum": ["NaN", "Infinity", "-Infinity"]}
+                      ]
+                    },
+                    "Z": { "type": ["string", "integer"] },
+                    "W" : { "type": "number" }
+                  }
+                }
+                """);
+
+            yield return new TestData<PocoWithRecursiveMembers>(
+                Value: new() { Value = 1, Next = new() { Value = 2, Next = new() { Value = 3 } } },
+                AdditionalValues: [new() { Value = 1, Next = null }],
+                ExpectedJsonSchema: """
+                {
+                    "type": ["object","null"],
+                    "properties": {
+                        "Value": { "type": "integer" },
+                        "Next": {
+                            "type": ["object", "null"],
+                            "properties": {
+                                "Value": { "type": "integer" },
+                                "Next": { "$ref": "#/properties/Next" }
+                            }
+                        }
+                    }
+                }
+                """);
+
+            // Same as above but using an anchor-based reference scheme
+            yield return new TestData<PocoWithRecursiveMembers>(
+                Value: new() { Value = 1, Next = new() { Value = 2, Next = new() { Value = 3 } } },
+                AdditionalValues: [new() { Value = 1, Next = null }],
+                ExpectedJsonSchema: """
+                {
+                    "$anchor" : "PocoWithRecursiveMembers",
+                    "type": ["object","null"],
+                    "properties": {
+                        "Value": { "type": "integer" },
+                        "Next": {
+                            "$anchor" : "PocoWithRecursiveMembers_Next",
+                            "type": ["object", "null"],
+                            "properties": {
+                                "Value": { "type": "integer" },
+                                "Next": { "$ref": "#PocoWithRecursiveMembers_Next" }
```

Comment:
FWIW the exporter itself never uses `$anchor` because I wanted to avoid having to unique generate identifiers for types. This particular tests validates that users can do that themselves via the provided callback mechanism.